clients/hadoopfs/src/main/java/io/lakefs/FileSystemTracer.java

```diff
@@ -0,0 +1,353 @@
+        }
+        return s3aRes;
+    }
+
+    @Override
+    public boolean delete(Path f, boolean recursive) throws IOException {
+        LOG.trace("delete(f {}, recursive {})", f, recursive);
+
+        boolean lakefsRes = lfsFileSystem.delete(f, recursive);
+        boolean s3aRes = s3AFileSystem.delete(translateLakeFSPathToS3APath(f), recursive);
+        LOG.trace("{}[delete] lakefs: {}, s3a: {}", RESULTS_COMPARISON, lakefsRes, s3aRes);
+
+        if (useLakeFSFileSystemRes) {
+            return lakefsRes;
+        }
+        return s3aRes;
+    }
+
+    @Override
+    public FileStatus[] listStatus(Path f) throws FileNotFoundException, IOException {
+        LOG.trace("listStatus(f {})", f);
+
+        FileStatus[] lakefsRes = lfsFileSystem.listStatus(f);
+        FileStatus[] s3aRes = s3AFileSystem.listStatus(translateLakeFSPathToS3APath(f));
+        LOG.trace("{}[listStatus] lakefs: {}, s3a: {}", RESULTS_COMPARISON, lakefsRes, s3aRes);
+
+        if (useLakeFSFileSystemRes) {
+            return lakefsRes;
+        }
+
+        for (FileStatus stat : s3aRes) {
+            Path s3aPath = stat.getPath();
+            Path lfsPath = translateS3APathToLakeFSPath(s3aPath);
+            stat.setPath(lfsPath);
+        }
+        return s3aRes;
+    }
+
+    /**
+     * Set the current working directory for the given file system. All relative
+     * paths will be resolved relative to it.
+     *
+     * @param new_dir
+     */
+    @Override
+    public void setWorkingDirectory(Path new_dir) {
+        LOG.trace("setWorkingDirectory(new_dir {})", new_dir);
+
+        if (useLakeFSFileSystemRes) {
+            lfsFileSystem.setWorkingDirectory(new_dir);
+        }
+        s3AFileSystem.setWorkingDirectory(translateLakeFSPathToS3APath(new_dir));
+    }
+
+    @Override
+    public Path getWorkingDirectory() {
+        LOG.trace("getWorkingDirectory()");
+
+        Path lakefsRes = lfsFileSystem.getWorkingDirectory();
+        Path s3aRes = s3AFileSystem.getWorkingDirectory();
+        LOG.trace("{}[getWorkingDirectory] lakefs: {}, s3a: {}", RESULTS_COMPARISON, lakefsRes, s3aRes);
+
+        if (useLakeFSFileSystemRes) {
+            return lakefsRes;
+        }
+        return s3aRes;
+    }
+
+    @Override
+    public boolean mkdirs(Path f, FsPermission permission) throws IOException {
+        LOG.trace("mkdirs(f {}, permission {})", f, permission);
+
+        boolean lakefsRes = lfsFileSystem.mkdirs(f, permission);
+        boolean s3aRes = s3AFileSystem.mkdirs(translateLakeFSPathToS3APath(f), permission);
+        LOG.trace("{}s[mkdirs] lakefs: {}, s3a: {}", RESULTS_COMPARISON, lakefsRes, s3aRes);
+
+        if (useLakeFSFileSystemRes) {
+            return lakefsRes;
+        }
+        return s3aRes;
+    }
+
+    @Override
+    public FileStatus getFileStatus(Path f) throws IOException {
+        LOG.trace("getFileStatus(f {})", f);
+
+        FileStatus lakefsRes = new FileStatus();
+        FileStatus s3aRes = new FileStatus();
```

Comment:
No need to allocate default values.
We need to capture the value returned and the exception.
if we have an exception we don't need to address the value, so we don't need the extra boolean too.